daemon/cmd/daemon_main.go

```diff
@@ -1719,16 +1719,44 @@ func newDaemonPromise(params daemonParams) (promise.Promise[*Daemon], promise.Pr
 			daemon = d
 
 			if !option.Config.DryMode {
-				if err := startDaemon(daemon, restoredEndpoints, cleaner, params); err != nil {
-					daemonResolver.Reject(err)
-					cancelDaemonCtx()
-					cleaner.Clean()
-					cfgResolver.Reject(err)
-					return err
+				log.Info("Initializing daemon")
+
+				// This validation needs to be done outside of the agent until
+				// datapath.NodeAddressing is used consistently across the code base.
+				log.Info("Validating configured node address ranges")
+				if err := node.ValidatePostInit(); err != nil {
+					return fmt.Errorf("postinit failed: %w", err)
+				}
+
+				err = option.Config.StoreInFile(option.Config.StateDir)
+				if err != nil {
+					log.WithError(err).Error("Unable to store Cilium's configuration")
+				}
+
+				err = option.StoreViperInFile(option.Config.StateDir)
+				if err != nil {
+					log.WithError(err).Error("Unable to store Viper's configuration")
 				}
 			}
-			daemonResolver.Resolve(daemon)
+
+			// NOTE: 'option.Config' must not be modified after this point
+			// (apparently startDaemon() does not modify option.Config)
 			cfgResolver.Resolve(option.Config)
+
+			if option.Config.DryMode {
+				daemonResolver.Resolve(daemon)
+			} else {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					if err := startDaemon(daemon, restoredEndpoints, cleaner, params); err != nil {
```

Comment:
Are we sure there are no start hooks which implicitly relied on `startDaemon` (and thereby endpoint regeneration) to be completed? In particular ones that are relying on `option.DaemonConfig`, which I believe we have a few.

Reply:
Not 100% sure, but it looks like it. No-one has access to the daemon before startDaemon completes as the promise is resolved... or at least that is how it should be :-)

Reply:
I don't think we can rely on type safety here unfortunately. For example, anything which has access to the EndpointManager and DaemonConfig was previously waiting for endpoint regeneration to finish. Now, that start hooks executed _before_ endpoint regeneration, which means they might observe an empty endpoint manager.

One doesn't need to have a reference to `*Daemon` in order to observe changes in behavior with this commit

Reply:
So maybe we need to add a promise for EndpointManager to fix this, or make the users of EndpointManager await for the restorer promise? Can you think of a systematical way we could find issues with this?

Reply:
So this concern only applies to cells that are dependent on `DaemonConfig`, as it is now resolved earlier than the Daemon promise - this narrows down the concern quite a bit: go through dependents of `DaemonConfig` and add an await for the Daemon promise when needed?

Or is this dependency transitive, if one cell depends on `DaemonConfig` and a later cell then depends on `EndpointManager`, then that cell may see an empty endpoint manager?

Reply:
There is only one dependency on`promise.Promise[*option.DaemonConfig]` (pkg/maps/nat), which is then providing `promise.Promise[NatMap4], promise.Promise[NatMap6]` for which there do not seem to be significant dependencies:
```
% git grep "Promise.*NatMap"      
pkg/datapath/fake/cells.go:             func() (promise.Promise[nat.NatMap4], promise.Promise[nat.NatMap6]) {
pkg/maps/nat/cell.go:   cell.Provide(func(lc cell.Lifecycle, cfgPromise promise.Promise[*option.DaemonConfig]) (promise.Promise[NatMap4], promise.Promise[NatMap6]) {
pkg/maps/nat/stats/stats.go:    NatMap4   promise.Promise[nat.NatMap4]
pkg/maps/nat/stats/stats.go:    NatMap6   promise.Promise[nat.NatMap6]
pkg/maps/nat/stats/stats_test.go:               cell.Provide(func() (promise.Promise[nat.NatMap4], promise.Promise[nat.NatMap6], Config, natMetrics) {
```

Reply:
Thanks for checking! Gives us much more confidence in the PR

Reply:
> There is only one dependency onpromise.Promise[*option.DaemonConfig] (pkg/maps/nat), which is then providing promise.Promise[NatMap4], promise.Promise[NatMap6] for which there do not seem to be significant dependencies:

For context, this is a recent change, this code depends on a stable `option.Config.EnableNodePort` which can change during the newDaemon/startDaemon phases. 

However there are likely other modules that may have potentially racey behavior due to the volatile nature of DaemonConfig.

Here's a issue I created for the problem: https://github.com/cilium/cilium/pull/32558

Reply:
Since currently there is just that one dependency on the DaemonConfig promise, could we just resolve it at the same time as we do daemonPromise below?

I think that should solve any issues.

Reply:
`finishKubeProxyReplacementInit()` is called from `newDaemon()`, before the DaemonConfig promise is resolved. AFAIK `option.Config` is not modified in `startDaemon()`.